File: experiments/go-replay/delivery.go
package main

import (
	"fmt"
	"math"
	"reflect"
	"sort"
	"time"
)

type Distribution struct {
	P50 float64 `json:"p50"`
	P95 float64 `json:"p95"`
	P99 float64 `json:"p99"`
	Max float64 `json:"max"`
}
type PhaseResult struct {
	FirstRecipientProgressMs Distribution        `json:"firstRecipientProgressMs"`
	MaximumRecipientLead     int                 `json:"maximumRecipientLead"`
	RecipientsWithProgress   int                 `json:"recipientsWithProgress"`
	QueueAgeP99Ms            float64             `json:"queueAgeP99Ms"`
	CpuMs                    float64             `json:"cpuMs"`
	QueueAgeOffsetMs         float64             `json:"queueAgeOffsetMs"`
	QueueAgeP50Ms            float64             `json:"queueAgeP50Ms"`
	QueueAgeP95Ms            float64             `json:"queueAgeP95Ms"`
	QueueAgeMaxMs            float64             `json:"queueAgeMaxMs"`
	ThroughputPerSecond      *float64            `json:"throughputPerSecond"`
	Name                     string              `json:"name"`
	Sent                     int                 `json:"sent"`
	Announcements            int                 `json:"announcements"`
	Retries                  int                 `json:"retries"`
	Attempts                 int                 `json:"attempts"`
	RecipientsAsserted       int                 `json:"recipientsAsserted"`
	ClassifiedRecipients     int                 `json:"classifiedRecipients"`
	DeliveriesByProfile      [][]string          `json:"deliveriesByProfile"`
	PayloadsByProfile        [][]Listing         `json:"payloadsByProfile"`
	ClassificationsByProfile []map[string]string `json:"classificationsByProfile"`
	ClassificationWallMs     float64             `json:"classificationWallMs"`
	WallMs                   float64             `json:"wallMs"`
	DrainMs                  float64             `json:"drainMs"`
	FirstProgressMaxMs       float64             `json:"firstProgressMaxMs"`
	MaxInFlight              int                 `json:"maxInFlight"`
	RateLimitsVerified       bool                `json:"rateLimitsVerified"`
}

// One event loop owns SQLite. Each turn starts at most one send per ready user;
// waits have deadlines, and only transport latency occupies one of eight slots.
func (s *Store) deliver(m Manifest, users int, catchup bool, mode string, now int64, out *PhaseResult) error {
	type attempt struct {
		at    float64
		retry bool
	}
	history := make([][]attempt, users)
	lastAttempt := -1e30
	type recipient struct {
		tokens, refilled, ready     float64
		announcement, retried, done bool
		attempts                    int
		sent                        []Listing
	}
	type flight struct {
		user    int
		listing *Listing
		end     float64
		retry   bool
		fail    bool
	}
	states := make([]recipient, users)
	for i := range states {
		states[i] = recipient{tokens: float64(m.Transport.RecipientBurst), announcement: catchup, sent: []Listing{}}
	}
	ages := []float64{}
	first := []float64{}
	progressCounts := make([]int, 1)
	progressCounts[0] = users
	minimumProgress, maximumProgress := 0, 0
	start := time.Now()
	clock := 0.0
	global := 0.0
	cursor := 0
	finished := 0
	flights := []flight{}
	for finished < users || len(flights) > 0 {
		if mode == "wall" {
			clock = float64(time.Since(start).Microseconds()) / 1000
		}
		pending := flights[:0]
		for _, f := range flights {
			if f.end > clock {
				pending = append(pending, f)
				continue
			}
			r := &states[f.user]
			if f.fail {
				r.done = true
				finished++
				continue
			}
			if f.retry {
				r.retried = true
				r.ready = clock + float64(m.Transport.RetryAfterMs)
				out.Retries++
				continue
			}
			r.ready = clock
			r.attempts = 0
			if f.listing == nil {
				r.announcement = false
				out.Announcements++
			} else {
				if e := s.acknowledge(f.user, f.listing.ID, now); e != nil {
					return e
				}
				if len(r.sent) == 0 {
					progress := clock
					if mode == "wall" {
						progress += out.ClassificationWallMs
					}
					first = append(first, progress)
					if clock > out.FirstProgressMaxMs {
						out.FirstProgressMaxMs = clock
					}
				}
				progressCounts[len(r.sent)]--
				if len(progressCounts) == len(r.sent)+1 {
					progressCounts = append(progressCounts, 0)
				}
				progressCounts[len(r.sent)+1]++
				for progressCounts[minimumProgress] == 0 {
					minimumProgress++
				}
				if len(r.sent)+1 > maximumProgress {
					maximumProgress = len(r.sent) + 1
				}
				if maximumProgress-minimumProgress > out.MaximumRecipientLead {
					out.MaximumRecipientLead = maximumProgress - minimumProgress
				}
				r.sent = append(r.sent, *f.listing)
				out.Sent++
				age := clock + out.QueueAgeOffsetMs
				if mode == "wall" {
					age += out.ClassificationWallMs
				}
				ages = append(ages, age)
			}
		}
		flights = pending
		launched := false
		next := 1e30
		for _, f := range flights {
			if f.end < next {
				next = f.end
			}
		}
		for scanned := 0; scanned < users && len(flights) < 8; scanned++ {
			if clock < global {
				if global < next {
					next = global
				}
				break
			}
			u := cursor
			cursor = (cursor + 1) % users
			// A retry whose deadline has elapsed gets its first progress before another sweep.
			for candidate := 0; candidate < users; candidate++ {
				waiting := &states[candidate]
				if waiting.retried && !waiting.done && len(waiting.sent) == 0 && waiting.ready <= clock {
					u = candidate
					break
				}
			}
			r := &states[u]
			if r.done {
				continue
			}
			if r.ready > clock {
				if r.ready < next {
					next = r.ready
				}
				continue
			}
			l, e := s.next(u)
			if e != nil {
				return e
			}
			if l == nil && !r.announcement {
				r.done = true
				finished++
				continue
			}
			r.tokens += (clock - r.refilled) * float64(m.Transport.RecipientMessagesPerMinute) / 60000
			if r.tokens > float64(m.Transport.RecipientBurst) {
				r.tokens = float64(m.Transport.RecipientBurst)
			}
			r.refilled = clock
			// Fractional refills can land a few ulps below one at the deadline.
			if r.attempts == 0 && r.tokens+1e-9 < 1 {
				r.ready = clock + (1-r.tokens)*60000/float64(m.Transport.RecipientMessagesPerMinute)
				if r.ready < next {
					next = r.ready
				}
				continue
			}
			if clock-lastAttempt+0.000001 < 1000/float64(m.Transport.GlobalAttemptsPerSecond) {
				return fmt.Errorf("global transport rate exceeded")
			}
			lastAttempt = clock
			history[u] = append(history[u], attempt{clock, r.attempts > 0})
			if r.attempts == 0 {
				r.tokens--
				if r.tokens < 0 {
					r.tokens = 0
				}
			}
			r.attempts++
			if r.attempts > m.Transport.MaxAttempts {
				return fmt.Errorf("retry budget exceeded")
			}
			if r.announcement {
				l = nil
			}
			retry := catchup && l != nil && len(r.sent) == 0 && u%m.Transport.RetryRecipientsModulo == 0 && !r.retried
			end := clock + float64(m.Transport.LatencyMs)
			flights = append(flights, flight{u, l, end, retry, out.Name == "interrupted" && len(r.sent) == 2})
			r.ready = 1e30
			global = clock + 1000/float64(m.Transport.GlobalAttemptsPerSecond)
			out.Attempts++
			launched = true
			if len(flights) > out.MaxInFlight {
				out.MaxInFlight = len(flights)
			}
			break
		}
		if finished == users && len(flights) == 0 {
			break
		}
		if launched {
			continue
		}
		if next == 1e30 {
			return fmt.Errorf("scheduler deadlock")
		}
		if mode == "wall" {
			delay := next - float64(time.Since(start).Microseconds())/1000
			if delay > 0 {
				time.Sleep(time.Duration(delay * 1e6))
			}
		} else {
			clock = next
		}
	}
	out.DrainMs = clock
	out.FirstRecipientProgressMs = distribution(first)
	out.RecipientsWithProgress = len(first)
	ageDistribution := distribution(ages)
	out.QueueAgeP50Ms = ageDistribution.P50
	out.QueueAgeP95Ms = ageDistribution.P95
	out.QueueAgeP99Ms = ageDistribution.P99
	out.QueueAgeMaxMs = ageDistribution.Max
	for _, attempts := range history {
		logical := []float64{}
		for i, a := range attempts {
			if a.retry {
				if i == 0 || a.at-attempts[i-1].at+0.000001 < float64(m.Transport.LatencyMs+m.Transport.RetryAfterMs) {
					return fmt.Errorf("retry deadline violated")
				}
			} else {
				logical = append(logical, a.at)
			}
		}
		for i, first := range logical {
			for j := i; j < len(logical); j++ {
				allowed := float64(m.Transport.RecipientBurst) + (logical[j]-first)*float64(m.Transport.RecipientMessagesPerMinute)/60000
				if float64(j-i+1) > allowed+0.000001 {
					return fmt.Errorf("recipient transport rate exceeded")
				}
			}
		}
	}
	out.RateLimitsVerified = true
	out.DeliveriesByProfile = make([][]string, 4)
	out.PayloadsByProfile = make([][]Listing, 4)
	for u, r := range states {
		ids := []string{}
		for _, l := range r.sent {
			ids = append(ids, l.ID)
		}
		if u < 4 {
			out.DeliveriesByProfile[u] = ids
			out.PayloadsByProfile[u] = r.sent
		} else if !reflect.DeepEqual(r.sent, out.PayloadsByProfile[u%4]) {
			return fmt.Errorf("recipient %d payload/order differs", u)
		}
		out.RecipientsAsserted++
	}
	return nil
}

func distribution(values []float64) Distribution {
	sort.Float64s(values)
	if len(values) == 0 {
		return Distribution{}
	}
	quantile := func(q float64) float64 { return values[int(math.Ceil(float64(len(values))*q))-1] }
	return Distribution{quantile(.5), quantile(.95), quantile(.99), quantile(1)}
}

File: experiments/rust-replay/src/delivery.rs
use crate::{
    Result,
    model::{Listing, Manifest},
    store::Store,
};
use serde::Serialize;
use std::{
    collections::BTreeMap,
    time::{Duration, Instant},
};

#[derive(Default, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PhaseResult {
    pub name: String,
    pub sent: usize,
    pub announcements: usize,
    pub retries: usize,
    pub attempts: usize,
    pub recipients_asserted: usize,
    pub classified_recipients: usize,
    pub deliveries_by_profile: Vec<Vec<String>>,
    pub payloads_by_profile: Vec<Vec<Listing>>,
    pub classifications_by_profile: Vec<BTreeMap<String, String>>,
    pub classification_wall_ms: f64,
    pub wall_ms: f64,
    pub drain_ms: f64,
    pub first_progress_max_ms: f64,
    pub max_in_flight: usize,
    pub rate_limits_verified: bool,
}
struct Recipient {
    tokens: f64,
    refilled: f64,
    ready: f64,
    announcement: bool,
    retried: bool,
    done: bool,
    attempts: usize,
    sent: Vec<Listing>,
    history: Vec<(f64, bool)>,
}
struct Flight {
    user: usize,
    listing: Option<Listing>,
    end: f64,
    retry: bool,
}
pub fn elapsed(start: Instant) -> f64 {
    start.elapsed().as_secs_f64() * 1000.0
}

// Only simulated transport occupies slots; rate and retry waits retain deadlines.
pub fn deliver(
    store: &Store,
    m: &Manifest,
    users: usize,
    catchup: bool,
    wall: bool,
    now: i64,
    out: &mut PhaseResult,
) -> Result<()> {
    let t = &m.transport;
    let mut states: Vec<_> = (0..users)
        .map(|_| Recipient {
            tokens: t.recipient_burst,
            refilled: 0.0,
            ready: 0.0,
            announcement: catchup,
            retried: false,
            done: false,
            attempts: 0,
            sent: Vec::new(),
            history: Vec::new(),
        })
        .collect();
    let start = Instant::now();
    let (mut clock, mut global, mut last_attempt) = (0.0, 0.0, f64::NEG_INFINITY);
    let (mut cursor, mut finished) = (0, 0);
    let mut flights: Vec<Flight> = Vec::new();
    while finished < users || !flights.is_empty() {
        if wall {
            clock = elapsed(start);
        }
        let mut pending = Vec::new();
        for f in flights.drain(..) {
            if f.end > clock {
                pending.push(f);
                continue;
            }
            let r = &mut states[f.user];
            if f.retry {
                r.retried = true;
                r.ready = clock + t.retry_after_ms;
                out.retries += 1;
                continue;
            }
            r.ready = clock;
            r.attempts = 0;
            if let Some(l) = f.listing {
                store.acknowledge(f.user, &l.id, now)?;
                if r.sent.is_empty() {
                    out.first_progress_max_ms = out.first_progress_max_ms.max(clock);
                }
                r.sent.push(l);
                out.sent += 1;
            } else {
                r.announcement = false;
                out.announcements += 1;
            }
        }
        flights = pending;
        let mut next = flights.iter().map(|f| f.end).fold(f64::INFINITY, f64::min);
        let mut launched = false;
        for _ in 0..users {
            if flights.len() >= 8 {
                break;
            }
            if clock < global {
                next = next.min(global);
                break;
            }
            let user = cursor;
            cursor = (cursor + 1) % users;
            let r = &mut states[user];
            if r.done {
                continue;
            }
            if r.ready > clock {
                next = next.min(r.ready);
                continue;
            }
            let mut listing = store.next(user)?;
            if listing.is_none() && !r.announcement {
                r.done = true;
                finished += 1;
                continue;
            }
            r.tokens = (r.tokens
                + (clock - r.refilled) * t.recipient_messages_per_minute / 60000.0)
                .min(t.recipient_burst);
            r.refilled = clock;
            // Floating-point refill can land a few ulps below one at a deadline.
            if r.attempts == 0 && r.tokens + 1e-9 < 1.0 {
                r.ready = clock + (1.0 - r.tokens) * 60000.0 / t.recipient_messages_per_minute;
                next = next.min(r.ready);
                continue;
            }
            if clock - last_attempt + 1e-6 < 1000.0 / t.global_attempts_per_second {
                return Err("global rate exceeded".into());
            }
            last_attempt = clock;
            r.history.push((clock, r.attempts > 0));
            if r.attempts == 0 {
                r.tokens = (r.tokens - 1.0).max(0.0);
            }
            r.attempts += 1;
            if r.attempts > t.max_attempts {
                return Err("retry budget exceeded".into());
            }
            if r.announcement {
                listing = None;
            }
            let retry = catchup
                && listing.is_some()
                && r.sent.is_empty()
                && user % t.retry_recipients_modulo == 0
                && !r.retried;
            flights.push(Flight {
                user,
                listing,
                end: clock + t.latency_ms,
                retry,
            });
            r.ready = f64::INFINITY;
            global = clock + 1000.0 / t.global_attempts_per_second;
            out.attempts += 1;
            out.max_in_flight = out.max_in_flight.max(flights.len());
            launched = true;
            break;
        }
        if finished == users && flights.is_empty() {
            break;
        }
        if launched {
            continue;
        }
        if !next.is_finite() {
            return Err("scheduler deadlock".into());
        }
        if wall {
            let delay = next - elapsed(start);
            if delay > 0.0 {
                std::thread::sleep(Duration::from_secs_f64(delay / 1000.0));
            }
        } else {
            clock = next;
        }
    }
    out.drain_ms = clock;
    for (user, r) in states.into_iter().enumerate() {
        let mut logical = Vec::new();
        for (i, &(at, retry)) in r.history.iter().enumerate() {
            if retry {
                if i == 0 || at - r.history[i - 1].0 + 1e-6 < t.latency_ms + t.retry_after_ms {
                    return Err("retry deadline violated".into());
                }
            } else {
                logical.push(at);
            }
        }
        for (i, first) in logical.iter().enumerate() {
            for (j, last) in logical.iter().enumerate().skip(i) {
                if (j - i + 1) as f64
                    > t.recipient_burst
                        + (last - first) * t.recipient_messages_per_minute / 60000.0
                        + 1e-6
                {
                    return Err("recipient rate exceeded".into());
                }
            }
        }
        if user < 4 {
            out.deliveries_by_profile
                .push(r.sent.iter().map(|l| l.id.clone()).collect());
            out.payloads_by_profile.push(r.sent);
        } else if r.sent != out.payloads_by_profile[user % 4] {
            return Err(format!("recipient {user} payload/order differs").into());
        }
        out.recipients_asserted += 1;
    }
    out.rate_limits_verified = true;
    Ok(())
}
pub fn observe(store: &Store, users: usize, ids: &[String], out: &mut PhaseResult) -> Result<()> {
    out.classifications_by_profile.clear();
    out.classified_recipients = 0;
    for user in 0..users {
        let decisions = store.classifications(user, ids)?;
        if user < 4 {
            out.classifications_by_profile.push(decisions);
        } else if decisions != out.classifications_by_profile[user % 4] {
            return Err(format!("recipient {user} classifications differ").into());
        }
        out.classified_recipients += 1;
    }
    Ok(())
}
